rq/batch.py

```diff
@@ -0,0 +1,96 @@
+from typing import List, Optional
+from uuid import uuid4
+
+from redis import Redis
+from redis.client import Pipeline
+
+from .exceptions import NoSuchBatchError
+from .job import Job
+from .utils import as_text
+
+
+class Batch:
+    """A Batch is a container for tracking multiple jobs with a single identifier."""
+
+    REDIS_BATCH_NAME_PREFIX = 'rq:batch:'
+    REDIS_BATCH_KEY = 'rq:batches'
+
+    def __init__(self, connection: Redis, id: str = None, jobs: List[Job] = None):
+        self.id = id if id else str(uuid4())
+        self.connection = connection
+        self.key = '{0}{1}'.format(self.REDIS_BATCH_NAME_PREFIX, self.id)
+
+        self.jobs = []
+        if jobs:
+            with connection.pipeline() as pipeline:
+                self.add_jobs(jobs, pipeline)
+                pipeline.execute()
+
+    def __repr__(self):
+        return "Batch(id={}, {} job(s))".format(self.id, len(self.jobs))
+
+    def add_jobs(self, jobs: List[Job], pipeline: Optional['Pipeline'] = None):
+        """Add jobs to the batch"""
+        pipe = pipeline if pipeline else self.connection.pipeline()
+        pipe.sadd(self.key, *[job.id for job in jobs])
+        pipe.sadd(self.REDIS_BATCH_KEY, self.id)
+        self.jobs += jobs
+        for job in jobs:
+            job.set_batch_id(self.id, pipeline=pipe)
+            job.save(pipeline=pipe)
+        if pipeline is None:
+            pipe.execute()
+
+    def delete_expired_jobs(self, pipeline: Optional['Pipeline'] = None):
+        """Delete jobs from the batch's job registry that have been deleted or expired from Redis.
+        We assume while running this that alive jobs have all been fetched from Redis in fetch_jobs method"""
+
+        pipe = pipeline if pipeline else self.connection.pipeline()
+        job_ids = {as_text(job) for job in self.connection.smembers(self.key)}
+        expired_jobs = job_ids - set([job.id for job in self.jobs if job])  # Return jobs that can't be fetched
+        for job in expired_jobs:
+            pipe.srem(self.key, job)
+        if pipeline is None:
+            pipe.execute()
+
+    def fetch_jobs(self) -> list:
+        """Retrieve list of job IDs from the batch key in Redis, then fetch and add those jobs to attribute"""
+        job_ids = [as_text(job) for job in self.connection.smembers(self.key)]
+        self.jobs = [job for job in Job.fetch_many(job_ids, self.connection) if job is not None]
+
+    def refresh(self, pipeline: Optional['Pipeline'] = None):
```

Comment:
We shouldn't need to have `batch.refresh()` as we should never implicitly load jobs from Redis, this could be a very expensive operation if a batch contains a large number of jobs.